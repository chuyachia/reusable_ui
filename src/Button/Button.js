import React from "react";
import PropTypes from "prop-types";

const Button = ({ children, onClick, className, style }) => (
  <button className={className} onClick={onClick} style={style}>
    {children}
  </button>
);

Button.propTypes = {
  className: PropTypes.string,
  variant: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  hollow: PropTypes.bool,
  style: PropTypes.object,
};

export default Button;
