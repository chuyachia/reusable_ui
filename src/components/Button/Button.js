import React from "react";
import PropTypes from "prop-types";

const Button = props => <button {...props} />;

Button.propTypes = {
  variant: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  hollow: PropTypes.bool
};

export default Button;
