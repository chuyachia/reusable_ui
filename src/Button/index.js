import React from "react";
import styled from "@emotion/styled";

import useTheme from "../useTheme";
import Button from "./Button";
import ThemeContext from "../ThemeContext";

const ButtonContext = React.createContext("default");
const StyledButton = styled(Button)`
  cursor: ${props => (props.disabled ? "not-allowed" : "pointer")};
  background-color: ${props => {
    if (props.disabled) {
      return useTheme(props, "silver");
    }
    if (props.context === "input" || props.context === "select") {
      return "transparent";
    }
    if (props.hollow || props.minimal) {
      return useTheme(props, "white");
    }
    if (props.variant) {
      return useTheme(props, props.variant);
    }
    return useTheme(props, "primary");
  }};
  color: ${props => {
    if (props.context === "input" || props.context === "select") {
      return useTheme(props, "silver");
    }
    if (props.disabled) {
      return useTheme(props, "grey");
    }
    if (props.variant) {
      return props.hollow || props.disabled
        ? useTheme(props, props.variant)
        : useTheme(props, props.variant + "ContrastText");
    }
    return props.hollow
      ? useTheme(props, "primary")
      : useTheme(props, "primaryContrastText");
  }};
  font-family: ${props => useTheme(props, "fontSans")};
  padding: ${props => {
    if (props.context === "input" || props.context === "select") {
      return 0;
    } else if (props.small) {
      return useTheme(props, "smallPadding");
    } else {
      return useTheme(props, "basePadding");
    }
  }};
  font-size: ${props => useTheme(props, "baseFontSize")};
  font-weight: ${props => useTheme(props, "fontSemibold")};
  line-height: ${props => {
    if (props.small) {
      return useTheme(props, "smallLineHeight");
    } else {
      return useTheme(props, "baseLineHeight");
    }
  }};
  border-width: ${props => {
    return useTheme(props, "borderWidth");
  }};
  border-style: solid;
  border-color: ${props => {
    if (props.disabled) {
      return useTheme(props, "silver");
    }
    if (props.variant) {
      return useTheme(props, props.variant);
    }
    return useTheme(props, "primary");
  }};
  border-radius: ${props => useTheme(props, "baseRadius")};
  ${props => {
    if (
      props.minimal ||
      props.context === "select" ||
      props.context === "input"
    ) {
      return "box-shadow: 0px;";
    } else {
      return "box-shadow:" + useTheme(props, "baseShadow") + ";";
    }
  }}}
  &:hover {
    background-color: ${props => {
      if (
        props.minimal ||
        props.context === "select" ||
        props.context === "input"
      ) {
        return useTheme(props, "white");
      }
      if (props.disabled) {
        return useTheme(props, "silver");
      }
      if (props.variant) {
        return props.hollow
          ? useTheme(props, props.variant + "Light")
          : useTheme(props, props.variant + "Dark");
      }
      return props.hollow
        ? useTheme(props, "primaryLight")
        : useTheme(props, "primaryDark");
    }};
    ${props => {
      if (props.context === "input" || props.context === "select") {
        return "color:" + useTheme(props, "baseTextColor");
      }
    }};
  }
  &:active {
    box-shadow: 0px 0px 0px;
  }
  ${props => {
    if (props.context === "input" || props.context === "select") {
      return "border:0;";
    }
  }};
  ${props => {
    if (props.context == "select") {
      return "float:right;";
    }
  }};
`;
export { ButtonContext };
const StyledButtonWContext = props => (
  <ButtonContext.Consumer>
    {context => (
      <ThemeContext.Consumer>
        {({ theme }) => (
          <StyledButton {...props} context={context} theme={theme} />
        )}
      </ThemeContext.Consumer>
    )}
  </ButtonContext.Consumer>
);

export default StyledButtonWContext;
