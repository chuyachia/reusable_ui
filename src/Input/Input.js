import React from "react";
import PropTypes from "prop-types";

import Button from "../Button";

const Input = ({
  resetButton,
  onReset,
  onKeyDown,
  value,
  controlled,
  inline,
  ...props
}) => {
  var inputProps = {
    onKeyDown,
    onChange: () => null,
  };
  if (controlled) inputProps.value = value;

  return (
    <div {...props}>
      <input {...inputProps} />
      {resetButton && <Button onClick={onReset}>X</Button>}
    </div>
  );
};

Input.propTypes = {
  controlled: PropTypes.bool,
  onKeyDown: PropTypes.func,
  onFocus: PropTypes.func,
  onChange: PropTypes.func,
  onReset: PropTypes.func,
  resetButton: PropTypes.bool,
  value: PropTypes.string,
  inline: PropTypes.bool,
};

export default Input;
