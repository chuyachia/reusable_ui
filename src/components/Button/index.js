import React from "react";
import styled from "@emotion/styled";

import Button from "./Button";

const ButtonContext = React.createContext("default");
const StyledButton = styled(Button)`
  cursor: pointer;
  background-color: ${props => {
    if (props.hollow) {
      return props.theme.white;
    } else if (props.context === "input" || props.context === "select") {
      return "transparent";
    } else {
      switch (props.variant) {
        case "secondary":
          return props.theme.secondary;
        case "warning":
          return props.theme.warning;
        default:
          return props.theme.primary;
      }
    }
  }};
  color: ${props => {
    if (props.context === "input" || props.context === "select") {
      return props.theme.baseTextColor;
    }
    switch (props.variant) {
      case "secondary":
        return props.hollow
          ? props.theme.secondary
          : props.theme.secondaryContrastText;
      case "warning":
        return props.hollow
          ? props.theme.warning
          : props.theme.warningContrastText;
      default:
        return props.hollow
          ? props.theme.primary
          : props.theme.primaryContrastText;
    }
  }};
  font-family: ${props => props.theme.fontSans};
  padding: ${props => {
    if (props.context === "input" || props.context === "select") {
      return 0;
    } else {
      return props.theme.basePadding;
    }
  }};
  font-size: ${props => props.theme.baseFontSize};
  font-weight: ${props => props.theme.fontSemibold};
  line-height: ${props => props.theme.baseLineHeight};
  outline: 0;
  border-width: ${props => props.theme.borderWidth};
  border-style: solid;
  border-color: ${props => {
    switch (props.variant) {
      case "secondary":
        return props.theme.secondary;
      case "warning":
        return props.theme.warning;
      default:
        return props.theme.primary;
    }
  }};
  border-radius: ${props => props.theme.baseRadius};
  &:hover {
    ${props => {
      if (props.hollow) {
        switch (props.variant) {
          case "secondary":
            return "background-color:" + props.theme.secondaryLight;
          case "warning":
            return "background-color:" + props.theme.warningLight;
          default:
            return "background-color:" + props.theme.primaryLight;
        }
      } else if (props.context === "input" || props.context === "select") {
        return "font-weight:" + props.theme.fontBold;
      } else {
        switch (props.variant) {
          case "secondary":
            return "background-color:" + props.theme.secondaryDark;
          case "warning":
            return "background-color:" + props.theme.warningDark;
          default:
            return "background-color:" + props.theme.primaryDark;
        }
      }
    }};
  }

  ${props => {
    if (props.context !== "default") {
      return "border:0;";
    }
  }};
  ${props => {
    if (props.context == "select") {
      return "float:right;";
    }
  }};
`;
export { ButtonContext };
export default props => (
  <ButtonContext.Consumer>
    {context => <StyledButton {...props} context={context} />}
  </ButtonContext.Consumer>
);
