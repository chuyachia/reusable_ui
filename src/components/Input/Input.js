import React from "react";
import PropTypes from "prop-types";

import Button from "../Button";

const Input = ({ resetButton, onReset, onChange, ...props }) => (
  <div {...props}>
    <input onChange={onChange}/>
    {resetButton && (
      <Button onClick={onReset}>
        X
      </Button>
    )}
  </div>
);

Input.propTypes = {
  onChange: PropTypes.func,
  onReset: PropTypes.func,
  resetButton: PropTypes.bool,
  value: PropTypes.string
};

export default Input;
