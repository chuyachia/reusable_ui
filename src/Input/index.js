import React from "react";
import styled from "@emotion/styled";

import useTheme from "../useTheme";
import Input from "./Input";
import { ButtonContext } from "../Button";

const StyledInput = styled(Input)`
  display: ${props => (props.inline ? "inline-block" : "block")};
  white-space: nowrap;
  color: ${props => useTheme(props, "baseTextColor")};
  font-weight: ${props => useTheme(props, "fontNormal")};
  line-height: ${props => useTheme(props, "baseLineHeight")};
  padding: ${props => useTheme(props, "basePadding")};
  font-family: ${props => useTheme(props, "fontSans")};
  font-size: ${props => useTheme(props, "baseFontSize")};
  border-width: ${props => useTheme(props, "borderWidth")};
  border-style: solid;
  border-color: ${props => {
    if (props.variant === "warning") {
      return useTheme(props, "warning");
    } else {
      return useTheme(props, "primary");
    }
  }};
  border-radius: ${props => useTheme(props, "baseRadius")};
  & input {
    width: 100%;
    border-width: 0px;
    padding: 0px;
    line-height: inherit;
    outline: 0;
    font-weight: ${props => useTheme(props, "fontNormal")};
    font-size: ${props => useTheme(props, "baseFontSize")};
    font-family: ${props => useTheme(props, "fontSans")};
  }
`;
const StyledInputWContext = props => (
  <ButtonContext.Provider value="input">
    <StyledInput {...props} />
  </ButtonContext.Provider>
);
export default StyledInputWContext;
