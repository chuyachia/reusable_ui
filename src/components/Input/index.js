import React from "react";
import styled from "@emotion/styled";

import Input from "./Input";
import { ButtonContext } from "../Button";

const StyledInput = styled(Input)`
  display: inline-block;
  white-space: nowrap;
  color: ${props => props.theme.baseTextColor};
  font-weight: ${props => props.theme.fontNormal};
  line-height: ${props => props.theme.baseLineHeight};
  padding: ${props => props.theme.basePadding};
  font-family: ${props => props.theme.fontSans};
  font-size: ${props => props.theme.baseFontSize};
  border-width: ${props => props.theme.borderWidth};
  border-style: solid;
  border-color: ${props => props.theme.primary};
  border-radius: ${props => props.theme.baseRadius};
  & input {
    border-width: 0px;
    padding: 0px;
    line-height: inherit;
    outline: 0;
    font-weight: ${props => props.theme.fontNormal};
    font-size: ${props => props.theme.baseFontSize};
    font-family: ${props => props.theme.fontSans};
  }
`;
export default props => (
  <ButtonContext.Provider value="input">
    <StyledInput {...props} />
  </ButtonContext.Provider>
);
