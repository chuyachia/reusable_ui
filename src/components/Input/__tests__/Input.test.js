import React from "react";
import Input from "../Input";
import Button from "../../Button";

describe("Input", () => {
  let wrapper;
  let mockFunc;
  it("should render", () => {
    wrapper = shallow(<Input />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render value property if controlled", () => {
    wrapper = shallow(<Input controlled value="A"/>);
    expect(wrapper.find('input').prop('value')).toBe("A");
  });
  it("should not render value property if not controlled", () => {
    wrapper = shallow(<Input value="A"/>);
    expect(wrapper.find('input').prop('value')).not.toBeDefined();
  });
  it("should renders reset button when resetButton is true", () => {
    wrapper = shallow(<Input resetButton={true} />);
    expect(wrapper.find(Button).length).toBe(1);
  });
  it("should call onChange function when recieving input value", () => {
    mockFunc = jest.fn();
    wrapper = shallow(
      <Input onChange={mockFunc} />
    );
    let input = wrapper.find("input");
    let changeEvent = { target: { value: "Hello" } };
    input.simulate("change", changeEvent);
    // expect(mockFunc.mock.calls[0][0]).toBe(changeEvent);
    // I thought mockFunc would match only the value
    expect(mockFunc).toBeCalledWith(changeEvent);
  });
  it("should call onKeyDown function when key pressed", () => {
    mockFunc = jest.fn();
    wrapper = shallow(
      <Input onKeyDown={mockFunc} />
    );
    let input = wrapper.find("input");
    input.simulate("keyDown");
    expect(mockFunc).toBeCalledTimes(1);
  });
  it("should call onFocus function when focused", () => {
    mockFunc = jest.fn();
    wrapper = shallow(
      <Input onFocus={mockFunc} />
    );
    let input = wrapper.find("input");
    input.simulate("focus");
    expect(mockFunc).toBeCalledTimes(1);
  });
});
