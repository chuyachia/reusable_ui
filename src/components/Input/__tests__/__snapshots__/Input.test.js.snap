// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    onChange={[MockFunction]}
    onReset={[MockFunction]}
    resetButton={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[MockFunction]}
        />,
        <_default
          onClick={[MockFunction]}
        >
          X
        </_default>,
      ],
      "className": "input-wrap",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [MockFunction],
          "onFocus": undefined,
          "onKeyDown": undefined,
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "X",
          "onClick": [MockFunction],
        },
        "ref": null,
        "rendered": "X",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[MockFunction]}
          />,
          <_default
            onClick={[MockFunction]}
          >
            X
          </_default>,
        ],
        "className": "input-wrap",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [MockFunction],
            "onFocus": undefined,
            "onKeyDown": undefined,
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "X",
            "onClick": [MockFunction],
          },
          "ref": null,
          "rendered": "X",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
